### ISSUE TITLE
do not recommend environment variables for configuration

### ISSUE THREAD

**Issue by brown** [cli-guidelines/cli-guidelines#48]
Please consider describing the use of shell environment variables as a technique of last resort for configuring commands.  Environment variables are like global variables in a program.  Global state is seductively effective at first but causes all kinds of problems when a system becomes more complex.  I've had to figure out many crazy bugs that boiled down to some command changing its behavior based on an environment variable.  For instance, if CDPATH is set in the shell environment, the "cd" bash command may produce output, ruining code like:  "output=$(cd "$dir" && some-command)".  At the very least, strongly recommend that any environment variables a command reads should be namespaced to diminish the change of conflict, so command "foo" should only ever look at environment variables with prefix "FOO_".

**Comment by nikclayton** [comment]
Indeed. As well as the problems above, environment variables complicate debugging, as well as making reproducing problems more difficult -- writing up the reproduction example requires you to ensure that you have explicitly mentioned all environment variables that are set/not-set and the reproducer has to take care to make sure they're system matches.

Compare with the simplicity of "Copy/paste this command to reproduce the problem."

**Comment by ssbarnea** [comment]
I don't think that use of environment variables is bad, quite the opposite. Still, abusing them would be an anti-pattern. Having few, well defined is very useful as it allows users to alter running configuration without having to touch static config files that may be under source-control.

Environment variables are the best way to control the CI/CD specific behaviors, NO_COLOR being an example.

There is a very limited number of generic env vars that should not be prefixed by tool name, but all the others should use a prefix.

**Comment by Quelklef** [comment]
> any environment variables a command reads should be namespaced

This is the way to go IMO.

Unfortunately, prefixing is just a band-aid, as it assumes that no two pieces of software will have the same name. However, most (all?) unix environments live on this assumption, and it's proven to work reasonably well.
